src/persist/src/indexed/future.rs

```diff
@@ -162,13 +223,63 @@ impl<K: Data, V: Data> BlobFuture<K, V> {
         Ok(())
     }
 
-    /// Remove all batches containing only data strictly below the Future's time
+    /// Remove all batches containing only data strictly before the Future's time
     /// lower bound.
     fn evict(&mut self) {
         // TODO: actually physically free the old batches.
         let ts_lower = self.ts_lower.clone();
         self.batches.retain(|b| ts_lower.less_equal(&b.ts_upper));
     }
+
+    /// Create a new [BlobFutureBatch] from `batch` containing only the subset of
+    /// updates at or in advance of the Future's time lower bound.
+    ///
+    /// `batch` is assumed not be eligible for eviction at the time of this function
+    /// call, and to satisy all [BlobFutureBatch] invariants.
+    fn trim<L: Blob>(
+        &mut self,
+        batch: BlobFutureBatchMeta,
+        blob: &mut BlobCache<K, V, L>,
+    ) -> Result<BlobFutureBatchMeta, Error> {
+        // Sanity check that batch cannot be evicted
+        debug_assert!(self.ts_lower.less_equal(&batch.ts_upper));
+        let mut updates = vec![];
+
+        updates.extend(
+            blob.get_future_batch(&batch.key)?
+                .updates
+                .iter()
+                .filter(|(_, ts, _)| self.ts_lower.less_equal(ts))
+                .cloned(),
+        );
+        debug_assert!(!updates.is_empty());
+        let new_batch = BlobFutureBatch {
+            desc: batch.desc,
+            updates,
+        };
+
+        self.write_batch(new_batch, blob)
+    }
+
+    /// Take one step towards shrinking the representation of this future.
+    ///
+    /// Returns true if the trace was modified, false otherwise.
+    pub fn step<L: Blob>(&mut self, blob: &mut BlobCache<K, V, L>) -> Result<bool, Error> {
```

Comment:
Is this intentionally not hooked up to anything yet?

Reply:
yup, same with merge in trace. I'm going to wire them both up with a separate thread doing the io and merging